### ISSUE TITLE
GemmBatched segfault on CPU (GPU unknown)

### ISSUE THREAD

**Issue by Ulfgard** [CNugteren/CLBlast#167]
Hi, i am running GemmBatched with float precision on my intel cpu device. I get a segfault somewhere inside the computation routines. Small problem sizes work fine, so i am confident that this is not an user error. Since I am not sure what causes it i have a hard time narrowing it down. The following parameters make it fail:

(in order of appearance)
Layout::kRowMajor, Transpose::kNo, Transpose::kYes,
1, 1, 40
alphas: 261*[1]
offsets_a 261*[0], lda 2560
offsets_b 261*[0], ldb 160
betas: 261*[1]
offsets_c [0,1,2,...,260], ldc 261

where buffer sizes are chosen apropriately (i think, buffer_a should be size 2560, buffer_b size 261 and buffer_c 261). The parameters seem to be fine and i could backtrace the issue to the XgemmDirectBatchedTT kernel. In there i can not debug any more without additional tools. 

//Edit i am using python list notation. so x*[y] is a list of x times the value y

**Comment by CNugteren** [comment]
On what OS and OpenCL are you on? Also, can you reproduce this in the included tests of CLBlast, or do they run fine? Or is this the same system with issue #166?

**Comment by Ulfgard** [comment]
This machine is an Arch Linux with Intel OpenCL 1.2 (the same as in #152 , see [clinfo](https://github.com/CNugteren/CLBlast/files/995480/clinfo.txt) )

All tests run fine on this machine.

**Comment by CNugteren** [comment]
Do I understand it correctly that all offsets of A and B are set to 0? In that case all matrices will overlap. You'll have to set the offsets properly manually I am afraid. For example, in the tests I have something like this to get a tight fit:
```
  for (auto batch = size_t{0}; batch < batch_count; ++batch) {
    a_offsets[batch] = batch * PerBatchSizeA(args) + base_a_offset;
    b_offsets[batch] = batch * PerBatchSizeB(args) + base_b_offset;
    c_offsets[batch] = batch * PerBatchSizeC(args) + base_c_offset;
  }
```

**Comment by Ulfgard** [comment]
Hi, in my real application all B matrices are the same as i compute C_i = A_i * B, so all b-offsets are 0.
for the sake of simplicity of presentation of the issue, i set the offsets of A_i to 0 due to the fact that my generated offsets are quite complicated and the simple case above already crashes with large enough problem sizes.

Background: I am using GemmBatched to implement a convolution over a set of images. Every row of A_i is made up of a subrow of an image ( of size filter-width * number-of-channels) and there are partial overlaps between the different A_i. The documentation does not state that this is not allowed and i do not see a technical reason why this should not be the case(otherwise: why having an offset array in the first place?).

**Comment by Ulfgard** [comment]
The real application code in question is here:
https://github.com/Shark-ML/Remora/blob/d6241b5ef24aaf5bf9d9a1dcf66c0c16a8858ccf/include/remora/kernels/clBlast/conv2d.hpp

**Comment by CNugteren** [comment]
>  there are partial overlaps between the different A_i. The documentation does not state that this is not allowed and i do not see a technical reason why this should not be the case(otherwise: why having an offset array in the first place?).

Yes, of course, you are right: that's allowed of course. Your example just seemed a bit funny so I assumed you made a mistake there, but now I understand that it's just a simplified example. I'll try to see if I can reproduce it myself.

**Comment by CNugteren** [comment]
This might well be the same issue as #140: it might be an issue with the GEMM kernels themselves. I've created a small branch (`gemm_direct_bug`) in which I've added a sample program mimicing your use-case. I tested it myself and get a segfault on my Intel CPU indeed. On my GPU the routine completes normally. Could you verify (just to be sure) if this is indeed your use-case and if you also see a crash with this sample (see `samples/sgemm_batched.cpp`).

I already noticed that changing line 114 in `src/routines/levelx/xgemmbatch.cpp` from `true` to `false` 'solves' the issue: it uses a different method to do GEMM with entirely different kernels. So I will dig into the gemm-direct kernels next. I'll let you know when the bug is identified and/or solved.

**Comment by Ulfgard** [comment]
The sample is in line with my parameters above and does also fail on my intel device.

Thank you for investigating this!

**Comment by CNugteren** [comment]
I believe I might have found the issue: there was a missing part of an if-statement guarding against invalid access of the A and B matrices in specific cases.

Could you try again with the latest version of the `gemm_direct_bug` branch? At least on my system it now runs OK. Please do also check it within your bigger application and whether the results are also valid (not crashing is not sufficient of course).

**Comment by Ulfgard** [comment]
The example works and runs, so does my testsuite for the convolution. I would consider this fixed

**Comment by CNugteren** [comment]
Great! I've just pushed this change to master. I think this also fixes #140.
